### PR TITLE
[sfmTransform] Creation of auto mode

### PR DESCRIPTION
**Pull request by servantftechnicolor**
New sfmTransform Node

- Use cameras X axis to estimate the horizontal plane (and the horizon)
- Use GPS measurements (if available) to align the Z axis with the north
- Use GPS measurements (if available) to set the scale (otherwise scaled such that stddev=1)
- Use cameras center mean as frame origin

See https://github.com/alicevision/Meshroom/pull/1954

### CODE REVIEW

File: src/aliceVision/sfm/utils/alignment.cpp [hunk 1]
```diff
@@ -986,5 +988,184 @@ void getRotationNullifyX(Eigen::Matrix3d & out_R, const Eigen::Vector3d & pt)
...
+
+    //Make sure that gps  
+    double var = gpsCov.trace();
+    if (var < distVariance)
+    {
+        ALICEVISION_LOG_INFO("Scene is too small to use GPS information");
+        return;
+    }
+        
+    ALICEVISION_LOG_INFO("GPS point cloud scale :" << gpsCov.trace());
```
Comment (fabiencastan) [inline src/aliceVision/sfm/utils/alignment.cpp:1138]: "GPS point cloud scale: "
Stupid formatting, but avoid space before ":" and put one after in all the places.

File: src/aliceVision/sfm/utils/alignment.cpp [hunk 2]
```diff
@@ -986,5 +988,184 @@ void getRotationNullifyX(Eigen::Matrix3d & out_R, const Eigen::Vector3d & pt)
...
+        Vec3 centeredGps = pair.second - gpsCoordinatesSum;
+        gpsCov += centeredGps * centeredGps.transpose();
+    }
+
+    //Make sure that gps  
+    double var = gpsCov.trace();
+    if (var < distVariance)
+    {
+        ALICEVISION_LOG_INFO("Scene is too small to use GPS information");
+        return;
```
Comment (fabiencastan) [inline src/aliceVision/sfm/utils/alignment.cpp:1135]:
```suggestion
        ALICEVISION_LOG_INFO("Scene is too small to use GPS information.");
        ALICEVISION_LOG_INFO("GPS variance is " << var << ", minimal required GPS variance is " << var << ".");
        return;
```

File: src/aliceVision/sfm/utils/alignment.cpp [hunk 3]
```diff
@@ -986,5 +988,184 @@ void getRotationNullifyX(Eigen::Matrix3d & out_R, const Eigen::Vector3d & pt)
     out_R = Eigen::AngleAxisd(angle, Vec3(0,-1,0)).toRotationMatrix();
 }
 
+void computeCentersMean(const sfmData::SfMData& sfmData, Vec3 & center)
```
[the remark this answers is not in the document]
Comment (simogasp) [reply src/aliceVision/sfm/utils/alignment.cpp:991]: Why not
```suggestion
Vec3 computeCameraCentersMean(const sfmData::SfMData& sfmData)
```

File: src/aliceVision/sfm/utils/alignment.cpp [hunk 4]
```diff
@@ -986,5 +988,184 @@ void getRotationNullifyX(Eigen::Matrix3d & out_R, const Eigen::Vector3d & pt)
...
+    const auto & poses = sfmData.getPoses();
+
+    for (auto v : sfmData.getViews())
+    {
+        if (!sfmData.isPoseAndIntrinsicDefined(v.first))
+        {
+            continue;
+        }
+
+        IndexT poseId = v.second->getPoseId();
```
Comment (simogasp) [inline src/aliceVision/sfm/utils/alignment.cpp:1005]:
```suggestion
        const auto poseId = v.second->getPoseId();
```

File: src/aliceVision/sfm/utils/alignment.cpp [hunk 5]
```diff
@@ -986,5 +988,184 @@ void getRotationNullifyX(Eigen::Matrix3d & out_R, const Eigen::Vector3d & pt)
...
+        {
+            continue;
+        }
+
+        if (!v.second->hasGpsMetadata())
+        {
+            continue;
+        }
+
+        IndexT poseId = v.second->getPoseId();
```
Comment (simogasp) [inline src/aliceVision/sfm/utils/alignment.cpp:1092]:
```suggestion
        const auto poseId = v.second->getPoseId();
```

File: src/aliceVision/sfm/utils/alignment.cpp [hunk 6]
```diff
@@ -986,5 +988,184 @@ void getRotationNullifyX(Eigen::Matrix3d & out_R, const Eigen::Vector3d & pt)
...
+
+    Eigen::Matrix3d camCov = Eigen::Matrix3d::Zero();
+    Eigen::Matrix3d gpsCov = Eigen::Matrix3d::Zero();
+    for (const auto & pair : list_pairs)
+    {
+        Vec3 centeredCam = pair.first - camCoordinatesSum;
+        camCov += centeredCam * centeredCam.transpose();
+        Vec3 centeredGps = pair.second - gpsCoordinatesSum;
+        gpsCov += centeredGps * centeredGps.transpose();
+    }
```
Comment (simogasp) [inline src/aliceVision/sfm/utils/alignment.cpp:1128]: can be broken down into 2 functions like before, it will improve readability

**Review by simogasp — CHANGES_REQUESTED** [review]

**Review by fabiencastan — APPROVED** [review]

**Pull request merged**